### DISCUSSION TITLE
helpneeded with samtools install - linking error libstdcxx-ng

### DISCUSSION THREAD

**Discussion by hardingnj** [prefix-dev/pixi#594] [General]
Hi,

Just looking for some help installing a package that depends on libstdcxx-n

Running into a linking error when installing samtools. Might be as simple as not having a core package installed? Thanks!

```
 INFO resolvo::solver: ╤══ Install libstdcxx-ng=libstdcxx-ng=13.2.0=h7e041cc_3 at level 31 (required by krb5=krb5=1.17.2=h926e7f8_0)
 WARN rattler::package_cache: validation failed: failed to read 'index.json'
 WARN rattler::package_cache: validation failed: failed to read 'index.json'
 WARN rattler::package_cache: validation failed: failed to read 'index.json'
 WARN rattler::package_cache: validation failed: failed to read 'index.json'
 WARN rattler::package_cache: validation failed: failed to read 'index.json'
 WARN rattler::package_cache: validation failed: failed to read 'index.json'
  × failed to link 'lib/terminfo'
  ├─▶ could not open destination file for writing
  ╰─▶ Is a directory (os error 21)

```

```
pixi 0.10.0
pixi global install samtools -c bioconda -c conda-forge -vv
```

System is Ubuntu 20.04 5.15.0-91-generic

**Comment by hardingnj** [comment]
```
 pixi global install libstdcxx-ng -c conda-forge -c bioconda
  × could not find an executable entrypoint in package libstdcxx-ng 13.2.0 h7e041cc_3 from conda-forge, are you sure it exists?

```

**Reply by ruben-arts** [reply]
`pixi global install` currently isn't build to install libraries. That is why you'll get this error. If you want to have libc available you should do it from a pixi project.

**Reply by hardingnj** [reply]
ok- that's helpful thanks.

**Comment by wolfv** [comment]
Can you also add -c conda-forge ?

**Reply by hardingnj** [reply]
Thanks- not following: `pixi global install samtools -c bioconda -c conda-forge -vv`

Also contains `-c conda-forge`?

**Reply by wolfv** [reply]
Sorry, didn't read it properly on mobile. Will have to investigate more

**Comment by hadim** [comment]
I am having a similar error running `pixi install` into a dev containter:

```bash
$ pixi install
  × failed to link 'share/terminfo/2/2621A'
  ╰─▶ No such file or directory (os error 2)
```

**Reply by ruben-arts** [reply]
Pixi can not link between file systems. So could it be you have the project on the host file system and run pixi in the container?

**Reply by wolfv** [reply]
I think this could also be an issue with case-sensitive file systems. Iirc there are some files that share the same name in lower-and upper case.

**Reply by hadim** [reply]
In case it helps, I have been able to make a minimal example to reproduce:

```Dockerfile
FROM --platform=linux/amd64 ubuntu:latest

# Install curl and tar for installing pixi binary
RUN apt-get update && apt-get install -y curl tar git

# Download and install pixi
ENV PLATFORM=unknown-linux-musl
RUN DOWNLOAD_URL="https://github.com/prefix-dev/pixi/releases/latest/download/pixi-$(uname -m)-$PLATFORM.tar.gz" &&
    curl -SL "$DOWNLOAD_URL" | tar -xz -C "/usr/local/bin"
```

```bash
# Build
docker build -t bug_linking .

# WORKS - no volume mounted
docker run --rm -ti \
  bug_linking \
  bash -c "git clone --depth=1 https://github.com/mamba-org/rattler.git /workspace/repo && cd /workspace/repo && pixi install"

# FAILS - `failed to link 'share/terminfo/n/ncr260vt300wpp'` error
docker run --rm -ti \
  -v myvol:/workspace/ \
  bug_linking \
  bash -c "git clone --depth=1 https://github.com/mamba-org/rattler.git /workspace/repo && cd /workspace/repo && pixi install"
```

---

I am running it on a MacBook Pro (not yet tested on a Linux machine).

As you can see, `pixi install` and only fails when done in a mounted volume.

---

> Pixi can not link between file systems. So could it be you have the project on the host file system and run pixi in the container?

What does conda/micromamba do in this situation? simply copying the file?

**Reply by baszalmstra** [reply]
The reason this happens is because the macOS filesystem is case insensitive whereas linux is case sensitive. That specific file exists with different capatalization in the package. Thats the reason this fails on mounted macOS volumes.

**Reply by hadim** [reply]
Ok - that's annoying 😅

So there is no way I can make this work, I guess?

Without switching the macOS disks from APFS to APFS case-sensitive? Which is possible but requires a full reinstallation...

**Reply by wolfv** [reply]
We will have to fix this in a similar manner as we are working on fixing "clobbering" in rattler (https://github.com/mamba-org/rattler/issues/460)

**Reply by baszalmstra** [reply]
Im not sure if there is a proper way to fix this. The problem is with the following:

The package `ncurses` for `linux-64` contains the following two files:

- `share/terminfo/2/2621A` 
- `share/terminfo/2/2621a` 

These are written to the cache which is stored in a macOS mounted volume. The mounted volume is case-insensitive which means these two files will clobber each other *inside* the cache. So the cache will contain either the first or the latter file, not both. You should be able to verify this by looking at the cache directory and finding the extracted `ncurses` package. Is there indeed only one file or is my reasoning wrong?

Anyway, the linux-64 based image expects a case-sensitive filesystem and thus expects both files to exist. When it's trying to hard-link the file it only finds one of the files which then results in the `╰─▶ No such file or directory (os error 2)` error. This is a limitation of docker unfortunately.

Did you confirm that this does indeed work with `conda`/`mamba`? I'm very curious how they solved this issue because I think its an inherent problem with docker on macOS.

One way to work around this issue is to not store the cache on your local machine but instead store it in the image itself. By default the cache is stored in your user directory. What happens if you don't mount that part?

**Reply by baszalmstra** [reply]
I created a small PR in rattler to hopefully diagnose these things better: https://github.com/mamba-org/rattler/pull/461

**Reply by hadim** [reply]
I finally switched my macOS to case-sensitive. I did a few tests and can confirm that the issue occurs only when the pixi folder is mounted.

I am also not sure if this can be easily fixed since it seems to be a per-package situation and I can imagine some of them requiring case sensitivity.

Maybe the best fix would be to throw a meaningful error message: "your pixi folder seems to be case-insensitive while the package XXXX does not support case-insensitive partition"

**Comment by hardingnj** [comment]
Not sure I follow the link between my issue and @hadim 's?

I'm not running on macOS- and don't think my issue due to capitalization.

**Reply by hadim** [reply]
My bad, as I am not sure your issue is indeed related to mine.

**Comment by lucascolley** [comment]
I think we can close this? FWIW this works for me:

```console
~/sandbox via 🧚 v0.50.2 
❯ pixi init pixi594; cd pixi594; pixi workspace channel add bioconda; pixi add -p linux-64 libstdcxx-ng
✔ Created /Users/lucascolley/sandbox/pixi594/pixi.toml
✔ Added bioconda (https://conda.anaconda.org/bioconda/)
✔ Added libstdcxx-ng >=15.1.0,<16
Added these only for platform(s): linux-64
```

please shout if you would still like some help @hardingnj !

EDIT: likewise for `pixi global install samtools -c conda-forge -c bioconda` on my Ubuntu box

**Reply by hardingnj** [reply]
Happy for it to be closed 🔒 

Thanks for checking
